projects/core/koin-core/src/commonMain/kotlin/org/koin/core/instance/ScopedInstanceFactory.kt

```diff
@@ -42,20 +50,20 @@ class ScopedInstanceFactory<T>(beanDefinition: BeanDefinition<T>) :
         return if (values[context.scope.id] == null) {
             super.create(context)
         } else {
-            values[context.scope.id] ?: error("Scoped instance not found for ${context.scope.id} in $beanDefinition")
+            values[context.scope.id] ?: throw MissingScopeValueException("Factory.create - Scoped instance not found for ${context.scope.id} in $beanDefinition")
         }
     }
 
     override fun get(context: ResolutionContext): T {
         if (context.scope.scopeQualifier != beanDefinition.scopeQualifier) {
-            error("Wrong Scope: trying to open instance for ${context.scope.id} in $beanDefinition")
+            error("Wrong Scope qualifier: trying to open instance for ${context.scope.id} in $beanDefinition")
         }
         KoinPlatformTools.synchronized(this) {
-            if (!isCreated(context)) {
-                values[context.scope.id] = create(context)
+            if (!isCreated(context) && holdInstance) {
+                values[context.scope.id] = super.create(context)
             }
         }
-        return values[context.scope.id] ?: error("Scoped instance not found for ${context.scope.id} in $beanDefinition")
+        return values[context.scope.id] ?: throw MissingScopeValueException("Factory.get -Scoped instance not found for ${context.scope.id} in $beanDefinition")
```

Comment:
This is a breaking change. Make sure you increase major version before release

Reply:
not in that case, as with such new behavior we would have ISE and can't use `getOrNull`.  Here external behavior is the same for the user, "just" a side effect internally to handle that case

Reply:
Then this api, including exception, should be internal

Reply:
you are right, it is still an internal API. I can tag them @KoinInternalApi in 4.1 branch.